### PR TITLE
My Reports should display "No Results" when export of district / scho…

### PR DESCRIPTION
**Pull request by dsnyder-fairway**
…ol contains no results

### CODE REVIEW

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcExamRepository.java [hunk 1]
```diff
@@ -34,21 +36,28 @@
...
                         .build(),
-                PreparedStatement::execute
+                (ps) -> {
+                    final ResultSet resultSet = ps.executeQuery();
+                    if (resultSet instanceof ResultSetImpl) {
+                        // Headers are one row. So, we have results if we have more than one row.
+                        return ((ResultSetImpl) resultSet).getUpdateCount() > 1;
+                    }
+                    return false;
+                }
```
Comment (dsnyder-fairway) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcExamRepository.java]: this I am not proud of, but I was unable to find a way to get the update count from a "SELECT ... INTO OUTFILE" command

Comment (dsnyder-fairway) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcExamRepository.java]: also not sure how we should handle this outside of MySQL

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/FileDetail.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package org.opentestsystem.rdw.reporting.processor.model;
+
+import java.net.URI;
+
+public class FileDetail {
```
Comment (esotrope) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/FileDetail.java]: naming nits: FileDetail -> FileDetails, hasResults -> hasContent

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcExamRepository.java [hunk 3]
```diff
@@ -34,21 +36,28 @@
     }
 
     @Override
-    public String export(final PermissionScope permissionScope, final ExportExamReportRequest request, final String uriLocation) {
+    public FileDetail export(final PermissionScope permissionScope, final ExportExamReportRequest request, final String uriLocation) {
         final boolean isDestinationS3 = uriLocation.toLowerCase().startsWith("s3");
 
-        template.execute(
+        final boolean hasResults = template.execute(
                 exportExam.replace(ExportDestination, isDestinationS3 ? "S3" : ""),
```
Comment (esotrope) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcExamRepository.java]: so the underlying sql writes a file to s3 directly and we never knew if the sql returned results or not. Now we execute the query directly and check how many rows the query returned to know if it had content or not.  - did i gather that correctly?

Comment (dsnyder-fairway) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcExamRepository.java]: yes -- except this shouldn't matter if written to s3 or now

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcExamRepository.java (hunk 1 above)
Comment (agorina) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcExamRepository.java]: Just curious: have we tested it on Aurora? 

An alternative approach I could see is to run a `SELECT ... LIMIT 1` before exporting and check if it returns result.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/FileDetail.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    private String relativePath;
+    private boolean hasResults;
+    private URI uri;
```
Comment (malaffoon) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/FileDetail.java]: Sometimes the relativePath will be set and sometimes the URI? If true, that feels a lot like a union structure from C days and i think we should talk about refactoring.

File: report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/stream/ExportExamsTest.java [hunk 5]
```diff
@@ -71,7 +72,7 @@ public void itShouldUpdateReportLocationUriAfterExport() {
         final URI exportUri = URI.create("EXPORT/EXAMS/test.csv");
 
         when(exportService.exportExam(any(User.class), any(ExportExamReportRequest.class), anyInt()))
-                .thenAnswer(x -> exportUri);
+                .thenAnswer(x -> FileDetail.builder().relativePath("EXPORT/EXAMS/test.csv").hasResults(true).uri(exportUri).build());
```
Comment (malaffoon) [inline report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/stream/ExportExamsTest.java]: Having the literal hard-coded here kind of defeats the point of the const variable. I'd suggest either hard-coding both (yech) or adding another const variable, e.g.
```
    final String relativePath = "EXPORT/EXAMS.test.csv";
    final URI exportUri = URI.create(relativePath);
    ...
```
Of course, all this makes me reiterate my earlier comment that the FileDetail structure feels ... messy.

**Comment by wtritch** [comment]
After looking this over I'm *really* tempted to say it's time to get rid of the whole "SELECT INTO OUTFILE" implementation and instead return results from the DB and print them to a CSV via Java...

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/FileDetail.java (hunk 4 above)
Comment (dsnyder-fairway) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/FileDetail.java]: relative path is used first, then the URI is set. I can probably refactor this to just use the URI

**Comment by dsnyder-fairway** [comment]
@wtritch @esotrope I'll go along with that suggestion.

**Comment by dsnyder-fairway** [comment]
closing for now

**Comment by agorina** [comment]
Sorry, I just saw @wtritch's  and @esotrope's suggestion above and I would like to voice my concern about it. If I recall it correctly, the reason COPY options was chosen is because this export could be a few millions of records, and COPY outperforms significantly (we are talking minutes vs hours). Maybe something to test before we do this change. Or just leave it as is..

**Pull request closed without merging**